ruby/basics/hashes.md

```diff
@@ -1,51 +1,185 @@
+
+```ruby
+shoes["winter"] = nil
+# And now we have no footwear in the snow.
+```
+
+The fun way to delete data from a hash is to use the hash's `#delete` method, which provides the cool functionality of returning the value from the key-value pair was deleted from the hash. So:
+
+```ruby
+shoes.delete("summer")
+  => "sandals"
+shoes
+  => {} # Empty hash, now that we have gotten rid of our shoes.
+```
+
+
+## Methods
+
+Hashes respond to just about all the same methods as arrays do, since they're both part of Ruby's *Enumerable* class. Since hashes have a few more moving parts, though, they behave a little differently. For example, whereas `array.each { |element| puts element }` will iterate through an array and print out each element in the array, hashes take two variables in the block, like so:
+
+```ruby
+hash.each { |key, value| puts "#{key}: #{value}" }
+```
+
+(Careful not to get confused between the curly braces denoting a block and the curly braces denoting a hash. They're generally pretty easy to tell apart, but keep an eye out.)
+
+Another couple of useful methods for hashes are the `#keys` and `#values` methods, which do just what you think they do. Note that these methods return *arrays*.
+
+```ruby
+books = { "Infinite Jest" => "David Foster Wallace", "Into the Wild" => "Jon Krakauer" }
+books.keys
+  => ["Infinite Jest", "Into the Wild"]
+books.values
+  => ["David Foster Wallace", "Jon Krakauer"]
+```
+
+### Merging two hashes
+
+It'll happen every now and again that two hashes need to come together in holy union. Luckily, there's a method for that.
+
+```ruby
+hash1 = { "a" => 100, "b" => 200 }
+hash2 = { "b" => 254, "c" => 300 }
+hash1.merge(hash2)
+  => { "a" => 100, "b" => 254, "c" => 300 }
+```
+
+Note that the hash getting merged in (in this case, `hash2`), has precedence over the hash getting... uh, merged *at*, when both hashes share a key.
+
+For a list of methods that work on hashes, check out the [Ruby Docs](http://ruby-doc.org/core-2.1.1/Hash.html).
+
+## Best practices
+
+In this explanation, we mostly used strings for hash keys, but in the real world, you'll see symbols like `:this_guy` as keys significantly more often. Especially when you get into Rails. The reason for this is a little complicated, but here's the short version:
+
+Strings can be changed, so every time a string is called, Ruby has to store it in memory. They can pile up, especially when hashes start to get big, and you can tell that Ruby is treating them all as different objects (even if they're otherwise identical) because they all have different `object_id`s. Symbols, on the other hand, are basically strings that can't change, and only get stored in memory once, which is faster and easier on your computer, since Ruby can search hash objects by their `object_id`s. For a better explanation of this behavior, check out Stack Overflow over [here](http://stackoverflow.com/questions/8189416/why-use-symbols-as-hash-keys-in-ruby).
```

Comment:
We have been thinking about changing the first lesson "Variables, strings and numbers" to include a description of of symbols as it seems to fit best there. We could move this excellent description you've done of symbols over there.

I think this section should focus on explaining the differences in the hash rocker and symbol syntax and to use the symbol syntax when possible as it looks cleaner and is more performant. Whats your thoughts on this?